### PR TITLE
installer: check warning message when uninstall driver by msi

### PR DESCRIPTION
**Pull request by leidwang**
ID: 1387

Signed-off-by: Leidong Wang <leidwang@redhat.com>

### CODE REVIEW

**Comment by vivianQizhu** [comment]
@leidwang Please consider to abstract common part to utils_windows. Thanks.

**Comment by leidwang** [comment]
(1/1) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.win_virtio_driver_install_by_installer.driver_uninstall.by_msi.q35: STARTED
 (1/1) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.win_virtio_driver_install_by_installer.driver_uninstall.by_msi.q35: PASS (473.32 s)

**Comment by leidwang** [comment]
> @leidwang Please consider to abstract common part to utils_windows. Thanks.

Hi @vivianQizhu 

In fact, the uninstall driver common part is that they(uninstall via exe/msi) all need to execute a command (this command will call the autoit script), and the others are different. So I think we might no need to define a function for it.

Not sure if I got your point, we can discuss it offline if needed.Thanks!

**Comment by leidwang** [comment]
Hi @heywji Could you please help review this patch?Thanks!

**Comment by heywji** [comment]
LGTM
```
1-repeat1.Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.win_virtio_driver_install_by_installer.driver_uninstall.by_msi.q35 Finshed 2023-11-15 00:33:53 2023-11-15 00:44:05 PASS 612.843096
2-repeat2.Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.win_virtio_driver_install_by_installer.driver_uninstall.by_msi.q35 Finshed 2023-11-15 00:44:06 2023-11-15 00:54:17 PASS 611.369202
3-repeat3.Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.win_virtio_driver_install_by_installer.driver_uninstall.by_msi.q35 Finshed 2023-11-15 00:54:18 2023-11-15 01:04:21 PASS 602.795255
Summary:
Finshed=3, PASS=3
```

**Comment by leidwang** [comment]
Hi @vivianQizhu Could you please help to review it when you are free?Thanks.

**Comment by leidwang** [comment]
avocado-vt 3808 still under review

**Comment by leidwang** [comment]
**Tested this patch with avocado-vt 3808, test will run as expected.**

1.Run the case in specific virtio-win range, the test case can pass.
JOB LOG    : /root/avocado/job-results/job-2024-01-23T18.31-c73ece2/job.log
 (1/2) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: STARTED
 (1/2) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (991.80 s)
 (2/2) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.win_virtio_driver_install_by_installer.driver_uninstall.by_msi.q35: STARTED
 (2/2) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.win_virtio_driver_install_by_installer.driver_uninstall.by_msi.q35: PASS (467.37 s)

2.Run the case not in specific virtio-win range, the test case can cancel.
JOB LOG    : /root/avocado/job-results/job-2024-01-23T19.07-06f6fec/job.log
 (1/1) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.win_virtio_driver_install_by_installer.driver_uninstall.by_msi.q35: STARTED
 (1/1) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.win_virtio_driver_install_by_installer.driver_uninstall.by_msi.q35: CANCEL: Got virtio-win version:1.9.34.0, which is not in [1.9.35.0, ) (3.93 s)

3.Run the case with wrong virtio-win limitation, the test case report an error.
JOB LOG    : /root/avocado/job-results/job-2024-01-23T19.35-56f313f/job.log
 (1/1) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.win_virtio_driver_install_by_installer.driver_uninstall.by_msi.q35: STARTED
 (1/1) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.win_virtio_driver_install_by_installer.driver_uninstall.by_msi.q35: ERROR: The limitation for virtio-win is not suitable for the cdrom_virtio (3.30 s)

Hi @vivianQizhu Would you please review this PR and avocado-vt PR again?Thanks!

File: qemu/tests/win_virtio_driver_installer_uninstall.py [hunk 1]
```diff
@@ -62,42 +62,56 @@ def run(test, params, env):
...
+        time.sleep(30)
+        run_uninstall_cmd = utils_misc.set_winutils_letter(
+            session, params["run_uninstall_cmd"])
+        session.cmd(run_uninstall_cmd)
+        time.sleep(30)
+        s_check, o_check = session.cmd_status_output(installer_pkg_check_cmd)
+        if s_check == 0:
+            test.fail("Could not uninstall Virtio-win-guest-tools package "
+                      "in guest', detail: '%s'" % o_check)
+        session = vm.reboot(session)
```
Comment (vivianQizhu) [inline qemu/tests/win_virtio_driver_installer_uninstall.py:91]: Let's introduce a params `uninstall_method`:

```suggestion
    uninstall_method = params.get('uninstall_method', 'installer')
    error_context.context("Run uninstall test via %s" % uninstall_method,
                              test.log.info)
    vm.send_key('meta_l-d')
    time.sleep(30)
    run_uninstall_cmd = utils_misc.set_winutils_letter(
        session, params["run_uninstall_cmd"])
    session.cmd(run_uninstall_cmd)
    time.sleep(30)
    if uninstall_method == "MSI":
        check_warning_file = params["check_warning_file"]
        output = session.cmd_output(check_warning_file)
        if params["warning_message"] not in output:
            test.fail("Not found expected warning message, the output is %s" % output)
    else:
        s_check, o_check = session.cmd_status_output(installer_pkg_check_cmd)
        if s_check == 0:
            test.fail("Could not uninstall Virtio-win-guest-tools package "
                      "in guest', detail: '%s'" % o_check)
        session = vm.reboot(session)
```

Comment (vivianQizhu) [reply qemu/tests/win_virtio_driver_installer_uninstall.py:91]: Another question to the test step itself: so we totally ignore whether the uninstall is successful or not via MSI? @xiagao @heywji ?

Comment (leidwang) [reply qemu/tests/win_virtio_driver_installer_uninstall.py]: > Let's introduce a params `uninstall_method`:

Updated,thanks!

Comment (heywji) [reply qemu/tests/win_virtio_driver_installer_uninstall.py]: Hi @vivianQizhu, Let me explain our test steps. The development has completely fixed the problem with missing DNS and IP addresses, but there are still unknown issues with some other parts of the virtio-win module. The developer still keeps this warning message, so our test step is to check that msi has this warning message when uninstalling. We do not ensure msi uninstalled is successful or not.

Comment (vivianQizhu) [reply qemu/tests/win_virtio_driver_installer_uninstall.py]: @heywji I understand you don't need to check it in your test with netkvm, but let's confirm with @xiagao If we really want to ignore it.

Comment (xiagao) [reply qemu/tests/win_virtio_driver_installer_uninstall.py]: In my opinion, 'Checking warning info' is just a step in win_virtio_driver_install_by_installer/uninstall test case, no need to check the uninstall result as from the test steps, 'tips exist and you can close the window if you can see it.'

File: qemu/tests/cfg/win_virtio_driver_install_by_installer.cfg [hunk 2]
```diff
@@ -243,6 +243,15 @@
...
             type = win_virtio_driver_installer_uninstall
+            variants:
+                - by_installer:
+                - by_msi:
+                    required_virtio_win = [1.9.35.0, )
+                    uninstall_method = msi
+                    run_uninstall_cmd = 'WIN_UTILS:\AutoIt3_%PROCESSOR_ARCHITECTURE%.exe WIN_UTILS:\uninstall_via_msi.au3'
+                    warning_text_path = "C:\uninstall_warning.txt"
+                    check_warning_file = 'type ${warning_text_path}'
+                    warning_message = ''
```
Comment (vivianQizhu) [inline qemu/tests/cfg/win_virtio_driver_install_by_installer.cfg:254]: Should we set a default value for it?

File: qemu/tests/win_virtio_driver_installer_uninstall.py [hunk 3]
```diff
@@ -62,19 +62,25 @@ def run(test, params, env):
...
-    s_check, o_check = session.cmd_status_output(installer_pkg_check_cmd)
-    if s_check == 0:
-        test.fail("Could not uninstall Virtio-win-guest-tools package "
-                  "in guest', detail: '%s'" % o_check)
-    session = vm.reboot(session)
+    if uninstall_method == "msi":
+        check_warning_file = params["check_warning_file"]
+        output = session.cmd_output(check_warning_file)
+        if params["warning_message"] not in output:
+            test.fail("Not found expected warning message, the output is %s" % output)
```
Comment (vivianQizhu) [inline qemu/tests/win_virtio_driver_installer_uninstall.py:77]: There is no reboot step but all the other check steps will be went through, do we know that they can all pass without reboot? Or maybe you want to add a `return` here to skip those steps?

File: qemu/tests/win_virtio_driver_installer_uninstall.py [hunk 4]
```diff
@@ -62,19 +62,25 @@ def run(test, params, env):
...
 
-    error_context.context("Run virtio-win-guest-tools.exe uninstall test",
+    uninstall_method = params.get('uninstall_method', 'installer')
+    error_context.context("Run uninstall test via %s" % uninstall_method,
                           test.log.info)
     vm.send_key('meta_l-d')
     time.sleep(30)
-    run_uninstall_cmd = utils_misc.set_winutils_letter(
-        session, params["run_uninstall_cmd"])
+    run_uninstall_cmd = utils_misc.set_winutils_letter(session, params["run_uninstall_cmd"])
```
[the remark this answers is not in the document]
Comment (leidwang) [reply qemu/tests/win_virtio_driver_installer_uninstall.py:70]: Okay.

File: qemu/tests/cfg/win_virtio_driver_install_by_installer.cfg (hunk 2 above)
Comment (leidwang) [reply qemu/tests/cfg/win_virtio_driver_install_by_installer.cfg:254]: Hi @vivianQizhu 

Sure, I also prefer to set a expected warning message. 
like that:https://github.com/autotest/tp-qemu/commit/3685f30bfda7b2b8ab85d3522f08bc57b835d121

File: qemu/tests/win_virtio_driver_installer_uninstall.py (hunk 3 above)
Comment (leidwang) [reply qemu/tests/win_virtio_driver_installer_uninstall.py:78]: Updated the test steps, it the test run with msi, there is no need to check if the driver is uninstalled.Thanks!

 (1/2) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: STARTED
 (1/2) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (969.12 s)
 (2/2) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.win_virtio_driver_install_by_installer.driver_uninstall.by_msi.q35: STARTED
 (2/2) Host_RHEL.m9.u4.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.win_virtio_driver_install_by_installer.driver_uninstall.by_msi.q35: PASS (501.53 s)

**Review by vivianQizhu — APPROVED** [review]

**Pull request merged**